src/platform/packages/shared/kbn-esql-ast/src/definitions/utils/validation/function.ts

```diff
@@ -6,930 +6,317 @@
+   */
+  private validateArguments(): ESQLMessage[] {
+    if (!this.definition) {
+      return [];
     }
 
-    args.forEach((argument, index) => {
-      const parameter = getParamAtPosition(signature, index);
-      if ((!argument && parameter?.optional) || !parameter) {
-        // that's ok, just skip it
-        // the else case is already catched with the argument counts check
-        // few lines above
-        return;
-      }
+    // Begin validation
+    const A = getSignaturesWithMatchingArity(this.definition, this.fn);
 
-      // check every element of the argument (may be an array of elements, or may be a single element)
-      const hasMultipleElements = Array.isArray(argument);
-      const argElements = hasMultipleElements ? argument : [argument];
-      const singularType = unwrapArrayOneLevel(parameter.type);
-      const messagesFromAllArgElements = argElements.flatMap((arg) => {
-        return [
-          validateFunctionLiteralArg,
-          validateNestedFunctionArg,
-          validateFunctionColumnArg,
-          validateInlineCastArg,
-        ].flatMap((validateFn) => {
-          return validateFn(
-            fn,
-            arg,
-            {
-              ...parameter,
-              type: singularType,
-              constantOnly: forceConstantOnly || parameter.constantOnly,
-            },
-            context,
-            parentCommand
-          );
-        });
-      });
-
-      const shouldCollapseMessages = isArrayType(parameter.type as string) && hasMultipleElements;
-
-      failingSignature.push(
-        ...(shouldCollapseMessages
-          ? collapseWrongArgumentTypeMessages(
-              messagesFromAllArgElements,
-              argument,
-              fn.name,
-              parameter.type as string,
-              parentCommand,
-              context
-            )
-          : messagesFromAllArgElements)
-      );
-    });
-    if (failingSignature.length) {
-      failingSignatures.push(failingSignature);
+    if (!A.length) {
+      return [errors.noMatchingCallSignatures(this.fn, this.definition, this.argTypes)];
     }
-  }
 
-  if (failingSignatures.length && failingSignatures.length === relevantFuncSignatures.length) {
-    const failingSignatureOrderedByErrorCount = failingSignatures
-      .map((arr, index) => ({ index, count: arr.length }))
-      .sort((a, b) => a.count - b.count);
-    const indexForShortestFailingsignature = failingSignatureOrderedByErrorCount[0].index;
-    messages.push(...failingSignatures[indexForShortestFailingsignature]);
-  }
-  // This is due to a special case in enrich where an implicit assignment is possible
-  // so the AST needs to store an explicit "columnX = columnX" which duplicates the message
-  return uniqBy(messages, ({ location }) => `${location.min}-${location.max}`);
-}
+    const S = getSignaturesMatchingTypes(this.definition, this.argTypes, this.argLiteralsMask);
```

Comment:
NiT: `const S` , `const A`
I didn't know whether to add a comment here, so ignore it in case, but maybe a more descriptive name can be useful (and camelCase)